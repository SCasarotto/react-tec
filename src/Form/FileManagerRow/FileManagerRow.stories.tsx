import React from 'react';

import { Story, Meta } from '@storybook/react';

import { FileManagerRow, FileManagerRowProps } from './FileManagerRow';

export default {
  title: 'Form/FileManagerRow',
  component: FileManagerRow,
  argTypes: {},
} as Meta;

const Template: Story<FileManagerRowProps> = (args) => (
  <FileManagerRow {...args} />
);

export const Default = Template.bind({});
Default.args = {
  labelForKey: 'someUniqueString',
  fileSrcArray: [
    {
      uid: 'uid1',
      src:
        'https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1024px-React-icon.svg.png',
      filename: 'SomeFile',
      fileEnding: 'png',
      path: 'some/path/to/where/i/have/this/file',
    },
    {
      uid: 'uid2',
      src:
        'https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1024px-React-icon.svg.png',
      filename: 'SomeTotallyDifferentFile',
      fileEnding: 'png',
      path: 'some/path/to/where/i/have/this/file',
    },
    {
      uid: 'uid3',
      src:
        'https://upload.wikimedia.org/wikipedia/commons/thumb/a/a7/React-icon.svg/1024px-React-icon.svg.png',
      filename: 'NotTheSameFileAsThoseOtherOnes',
      fileEnding: 'png',
      path: 'some/path/to/where/i/have/this/file',
    },
  ],
  onUpload: async (files) => {
    console.log(files);
    Promise.resolve();
  },
  onRemove: async (data) => console.log(data),
};
