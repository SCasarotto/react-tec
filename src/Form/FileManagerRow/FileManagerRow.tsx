import React, { useState } from 'react';

import { FormRowCustomProps, FormRow } from '../FormRow';
import { InputProps } from '../Input';
import { Label, LabelCustomProps } from '../Label';
import {
  FileRowWrapper,
  FileWrapper,
  FileName,
  DownloadFileButton,
  DownloadImageButtonIcon,
  ClearFileButton,
  ClearFileButtonIcon,
  Input,
} from './styledComponents';

//FileManagerRow âœ…
export interface FileManagerRowSrcObject {
  uid: string;
  src: string;
  filename: string;
  fileEnding: string;
  path: string;
}
export interface FileManagerRowOnRemoveData {
  uid: string;
  path: string;
  index: number;
}
/**
 *
 * A component to manage uploading, downloading, and deleting files. This component contains a `Row`, `Label`, `FileInput` and some additional components.
 *
 * Note: This component is a bit complex and may shift in API over time.
 *
 */
export interface FileManagerRowProps
  extends FormRowCustomProps,
    LabelCustomProps,
    InputProps {
  labelForKey: string;
  fileSrcArray: Array<FileManagerRowSrcObject>;
  maxNumber?: number;
  filePattern?: RegExp;
  // multiple?: false
  // TODO: improve these types
  onRemove(data: FileManagerRowOnRemoveData): Promise<any>;
  onUpload(files: FileList): Promise<any>;
}
export const FileManagerRow = React.forwardRef<
  HTMLInputElement,
  FileManagerRowProps
>((props, ref) => {
  const [inputKey, setInputKey] = useState<string | number>('inputKey');

  const {
    rowSize,
    last,
    className = '',
    title,
    fileSrcArray,
    maxNumber = 0,
    accept,
    filePattern,
    disabled,
    onRemove,
    required,
    labelForKey,
    onUpload,
    ...rest
  } = props;

  const onFileChange = (fileList: FileList) => {
    onUpload(fileList)
      .then(() => {
        setInputKey(new Date().getTime());
      })
      .catch((e) => {
        console.log(e);
        setInputKey(new Date().getTime());
      });
  };

  let labelText = title;
  if (labelText && maxNumber > 1 && fileSrcArray) {
    labelText = `${title} (${fileSrcArray.length}/${maxNumber})`;
  }

  return (
    <FormRow
      rowSize={rowSize}
      last={last}
      className={`tec-fileManagerRow ${className}`}
    >
      {labelText && (
        <Label
          htmlFor={labelForKey}
          required={required}
          disabled={disabled}
          className='tec-fileManagerRowTitle'
        >
          {labelText}
        </Label>
      )}
      <FileRowWrapper className='tec-fileManagerRowWrapper'>
        {fileSrcArray?.map((imgSrcData, index) => {
          const { src, uid, filename, fileEnding, path } = imgSrcData;
          return (
            <FileWrapper
              key={uid}
              className='tec-fileManagerFileRow'
              disabled={disabled}
            >
              <FileName className='tec-fileManagerFileName' disabled={disabled}>
                {filename}.{fileEnding}
              </FileName>
              {src && (
                <DownloadFileButton
                  href={src}
                  target='_blank'
                  rel='noopener noreferrer'
                  key={uid}
                  download={filename}
                  className='tec-fileManagerDownloadButton'
                >
                  <DownloadImageButtonIcon className='tec-fileManagerDownloadIcon' />
                </DownloadFileButton>
              )}
              {!disabled && (
                <ClearFileButton
                  onClick={() => onRemove({ uid, path, index })}
                  className='tec-fileManagerClearButton'
                >
                  <ClearFileButtonIcon className='tec-fileManagerClearIcon' />
                </ClearFileButton>
              )}
            </FileWrapper>
          );
        })}
        {(maxNumber === 0 ||
          !fileSrcArray ||
          (fileSrcArray && fileSrcArray.length < maxNumber)) && (
          <Input
            {...rest}
            key={inputKey} //Used to reset the input
            id={labelForKey}
            onChange={onFileChange}
            disabled={disabled}
            required={required}
            // multiple={maxNumber === 0 || maxNumber > 1} //For simplicity making it a single file
            accept={accept}
            filePattern={filePattern}
            className='tec-fileManagerInput'
            hideClearButton
            ref={ref}
          />
        )}
      </FileRowWrapper>
    </FormRow>
  );
});

FileManagerRow.defaultProps = {
  accept: '.',
  filePattern: /./,
  // multiple: false, //Not implemented for multiple so defaulting for safety
};
