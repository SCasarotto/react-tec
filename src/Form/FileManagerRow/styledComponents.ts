import { FaFileDownload, FaRegWindowClose } from 'react-icons/fa';
import styled from 'styled-components';

import { Button } from '../Button';
import { FileInput } from '../FileInput';

export const FileRowWrapper = styled.div``;
export const FileWrapper = styled.div<{ disabled?: boolean }>`
  display: flex;
  align-items: center;
  padding: 5px;
  border: 1px solid
    ${(props) =>
      props.disabled ? props.theme.lighterGray : props.theme.lightGray};
  background-color: ${(props) => props.theme.white};
`;
export const FileName = styled.span<{ disabled?: boolean }>`
  flex: 1;
  padding-left: 5px;
  display: block;
  color: ${(props) =>
    props.disabled ? props.theme.gray : props.theme.darkerGray};
`;
export const DownloadFileButton = styled.a`
  width: auto;
  padding: 5px 10px;
  color: ${(props) => props.theme.darkGray};
  background-color: ${(props) => props.theme.white};
  border: 1px solid ${(props) => props.theme.lightGray};
  border-radius: 5px;
  display: flex;
  align-items: center;
  justify-content: center;

  transition: color 0.2s ease-in 0s, background-color 0.2s ease-in 0s,
    border 0.2s ease-in 0s;
  :active,
  :hover {
    background-color: ${(props) => props.theme.primary};
    color: ${(props) => props.theme.white};
    border: 1px solid ${(props) => props.theme.primary};
  }
`;
export const DownloadImageButtonIcon = styled(FaFileDownload)`
  color: inherit;
  display: block;
  font-size: 18px;
`;
export const ClearFileButton = styled(Button)`
  width: auto;
  padding: 5px 10px;
  margin-left: 5;
`;
export const ClearFileButtonIcon = styled(FaRegWindowClose)`
  color: inherit;
  display: block;
  font-size: 18px;
`;
export const Input = styled(FileInput)`
  margin-top: 10px;
`;
