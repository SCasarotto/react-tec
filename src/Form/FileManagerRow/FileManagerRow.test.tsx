import React from 'react';

import { render, waitFor } from '@testing-library/react';

import { FileManagerRow } from './FileManagerRow';

describe('FileManagerRow', () => {
  // Renders Without Error
  it('matches snapshot', () => {
    const { asFragment } = render(
      <FileManagerRow
        labelForKey='testFileManagerRow'
        fileSrcArray={[]}
        onRemove={async () => jest.fn()}
        onUpload={async () => jest.fn()}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // Class Name
  it('accepts custom className', () => {
    const { container } = render(
      <FileManagerRow
        className='myClassName'
        labelForKey='testFileManagerRow'
        fileSrcArray={[]}
        onRemove={async () => jest.fn()}
        onUpload={async () => jest.fn()}
      />,
    );
    expect(container.firstChild).toHaveClass('myClassName');
  });

  // Ref
  it('renders with ref', async () => {
    const ref: { current: HTMLInputElement | null } = { current: null };
    const { getByTestId } = render(
      <FileManagerRow
        labelForKey='testFileManagerRow'
        fileSrcArray={[]}
        onRemove={async () => jest.fn()}
        onUpload={async () => jest.fn()}
        ref={ref}
        data-testid='myFileInput'
      />,
    );
    const fileInput = getByTestId('myFileInput');
    await waitFor(() => expect(fileInput).not.toHaveFocus());
    ref.current?.focus();
    await waitFor(() => expect(fileInput).toHaveFocus());
  });
});
