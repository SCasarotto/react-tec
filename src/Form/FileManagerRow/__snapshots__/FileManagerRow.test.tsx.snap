// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileManagerRow matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: inline-block;
  vertical-align: top;
  margin-bottom: 10px;
  text-align: left;
  box-sizing: border-box;
  padding-right: 0px;
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  border-radius: 5px;
  padding-right: 5px;
  border: 1px solid;
  box-shadow: none;
  -webkit-transition: border-color 0.2s ease-in,box-shadow 0.2s ease-in;
  transition: border-color 0.2s ease-in,box-shadow 0.2s ease-in;
}

.c3:hover,
.c3:active {
  border: 1px solid;
  box-shadow: 0 0 0 1px inset;
}

.c4 {
  display: block;
  width: 100%;
  font-size: 14px;
  padding: 10px;
  cursor: pointer;
}

.c5 {
  margin-right: 10px;
  vertical-align: middle;
  font-size: 18px;
}

.c6 {
  display: none;
}

.c2 {
  margin-top: 10px;
}

<div
    class="c0 tec-formRow tec-fileManagerRow "
  >
    <div
      class="tec-fileManagerRowWrapper"
    >
      <div
        class="c1 tec-fileInput c2 tec-fileManagerInput"
      >
        <div
          class="c3 tec-fileInputInputWrapper"
        >
          <label
            class="c4 tec-fileInputLabel"
          >
            <span
              class="tec-fileInputCopy"
            >
              <svg
                class="c5 tec-fileInputIcon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 384 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm65.18 216.01H224v80c0 8.84-7.16 16-16 16h-32c-8.84 0-16-7.16-16-16v-80H94.82c-14.28 0-21.41-17.29-11.27-27.36l96.42-95.7c6.65-6.61 17.39-6.61 24.04 0l96.42 95.7c10.15 10.07 3.03 27.36-11.25 27.36zM377 105L279.1 7c-4.5-4.5-10.6-7-17-7H256v128h128v-6.1c0-6.3-2.5-12.4-7-16.9z"
                />
              </svg>
              Choose a file...
            </span>
            <input
              accept="."
              class="c6 tec-fileInputInput"
              id="testFileManagerRow"
              type="file"
            />
          </label>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
